test/hotspot/jtreg/runtime/Monitor/TestRecursiveLocking.java

```diff
@@ -0,0 +1,560 @@
+            Asserts.assertTrue(WB.isMonitorInflated(this));
+        }
+    }
+
+    static void usage() {
+        System.err.println();
+        System.err.println("Usage: java TestRecursiveLocking [n_secs]");
+        System.err.println("       java TestRecursiveLocking n_secs [mode]");
+        System.err.println();
+        System.err.println("where:");
+        System.err.println("    n_secs  ::= > 0");
+        System.err.println("            Default n_secs is 15.");
+        System.err.println("    mode    ::= 1 - outer and inner");
+        System.err.println("            ::= 2 - alternate A and B");
+        System.err.println("            Default mode is 1.");
+        System.exit(1);
+    }
+
+    public static void main(String... argv) throws Exception {
+        int mode = 2;
+        int n_secs = 30;
+
+        if (argv.length != 0 && argv.length != 1 && argv.length != 2) {
+            usage();
+        } else if (argv.length > 0) {
+            try {
+                n_secs = Integer.parseInt(argv[0]);
+                if (n_secs <= 0) {
+                    throw new NumberFormatException("Not > 0: '" + argv[0]
+                                                    + "'");
+                }
+            } catch (NumberFormatException nfe) {
+                System.err.println();
+                System.err.println(nfe);
+                System.err.println("ERROR: '" + argv[0]
+                                   + "': invalid n_secs value.");
+                usage();
+            }
+
+            if (argv.length > 1) {
+                try {
+                    mode = Integer.parseInt(argv[1]);
+                    if (mode != 1 && mode != 2) {
+                        throw new NumberFormatException("Not 1 -> 2: '"
+                                                        + argv[1] + "'");
+                    }
+                } catch (NumberFormatException nfe) {
+                    System.err.println();
+                    System.err.println(nfe);
+                    System.err.println("ERROR: '" + argv[1]
+                                       + "': invalid mode value.");
+                    usage();
+                }
+            }
+        }
+
+        System.out.println("INFO: LockingMode=" + flagLockingMode);
+        System.out.println("INFO: LockStackCapacity=" + constLockStackCapacity);
+        System.out.println("INFO: n_secs=" + n_secs);
+        System.out.println("INFO: mode=" + mode);
+
+        long loopCount = 0;
+        long endTime = System.currentTimeMillis() + n_secs * 1000;
+
+        syncThread.waitForStart();
+
+        while (System.currentTimeMillis() < endTime) {
+            loopCount++;
+            SynchronizedObject syncObj = new SynchronizedObject();
+            switch (mode) {
+            case 1:
+                syncObj.runOuterInnerTest();
+                break;
+
+            case 2:
+                syncObj.runAlternateABTest();
+                break;
+
```

Comment:
I was wondering if this test could be reduced to run both modes  in one execution, one after the other?

Reply:
The configurability was intentional so that specific modes could be exercised independently. Also adding additional modes in the future is/was planned.